### ISSUE TITLE
[Proposal] Add Lllama 3.1 support

### ISSUE THREAD

**Issue by ssuukk** [TransformerLensOrg/TransformerLens#691]
### Proposal 

Add Llama 3.1 support. Currently trying to load it fails with:

`ValueError: meta-llama/Meta-Llama-3.1-8B-Instruct not found. Valid official model names (excl aliases): `

### Motivation

Llamma 3.1 is new bestest model, they say it's smarter than Rey Palpatine.

- [X] I have checked that there is no similar [issue](https://github.com/TransformerLensOrg/Transformerlens/issues) in the repo (**required**)

**Comment by mylesgoose** [comment]
It works alredy with llama 3.1 you just have to put it into a folder with the name of the previous llama like so. MODEL_ID = "meta-llama/Meta-Llama-3-70B-Instruct"
NEW_MODEL_ID = "mylesgoose/Meta-Llama-3.1-70B-abliterated"
MODEL_TYPE = "meta-llama/Meta-Llama-3-70B-Instruct"

# Load model and tokenizer on CPU first to avoid memory issues
model = HookedTransformer.from_pretrained_no_processing(
    MODEL_TYPE,
    local_files_only=True,
    dtype=torch.bfloat16,
    default_padding_side='left',
    device="cpu"
)
tokenizer = AutoTokenizer.from_pretrained(MODEL_TYPE)
tokenizer.padding_side = 'left'
tokenizer.pad_token = tokenizer.eos_token or you can add the new model directly where you find the other models in that class. for exampel when you run the coee and it gives you an error. click on the url for the erro message and edit the page to incldue the details for the new model. so basically just copy and paste the4 cotnents of Meta-Llama-3.1-70B into the 3 folder and run from local files as above. https://huggingface.co/blog/mlabonne/abliteration

**Comment by mylesgoose** [comment]
[loading_from_pretrained.zip](https://github.com/user-attachments/files/16535742/loading_from_pretrained.zip)

**Comment by ssuukk** [comment]
Adding it to loading_from_pretrained.py fails with:

```
  File "E:\Projekty\_AI\abliteration\main.py", line 66, in <module>
    model = HookedTransformer.from_pretrained_no_processing(
  File "C:\Users\ssuuk\anaconda3\envs\rope2\lib\site-packages\transformer_lens\HookedTransformer.py", line 1331, in from_pretrained_no_processing
    return cls.from_pretrained(
  File "C:\Users\ssuuk\anaconda3\envs\rope2\lib\site-packages\transformer_lens\HookedTransformer.py", line 1243, in from_pretrained
    cfg = loading.get_pretrained_model_config(
  File "C:\Users\ssuuk\anaconda3\envs\rope2\lib\site-packages\transformer_lens\loading_from_pretrained.py", line 1454, in get_pretrained_model_config
    cfg_dict = convert_hf_model_config(official_model_name, **kwargs)
  File "C:\Users\ssuuk\anaconda3\envs\rope2\lib\site-packages\transformer_lens\loading_from_pretrained.py", line 1074, in convert_hf_model_config
    "d_model": hf_config.hidden_size,
UnboundLocalError: local variable 'hf_config' referenced before assignment
```

**Comment by mylesgoose** [comment]
if i look at my loading _from_pretrain.py file the lines do not match up to the errors your getting above. is it differnt on windows to linux? Try that file folder swap thing i showed above. and restore your loading..py file back to original. and load from directory as i explained above. first try to load the original version 3 model from disk and then if it laodes withc in that sme fodler the contents of llamam3.1 and try run your same script again. it 100% works to load a 3.1 llama model from a 3.0 llama folder on ubuntu linux. try in a wsl2 environemnt maybe if that fails. if you take a look at this hugginface repo you can see he has even used this script to make a 3.1 model via loading with hooked transfomers. and he did the directory trick also as he explained in his artical. I tried it and it worked.  https://huggingface.co/mlabonne/Meta-Llama-3.1-8B-Instruct-abliterated here is the code i can run wihtout any modifications to the py file to add new models. jsut swappign the new model to the old models fodler and loading form lcoal not form huggin face cache. MODEL_ID = "meta-llama/Meta-Llama-3-70B-Instruct"
NEW_MODEL_ID = "mylesgoose/Meta-Llama-3.1-70B-abliterated"
MODEL_TYPE = "meta-llama/Meta-Llama-3-70B-Instruct"

# Load model and tokenizer on CPU first to avoid memory issues
model = HookedTransformer.from_pretrained_no_processing(
    MODEL_TYPE,
    local_files_only=True,
    dtype=torch.bfloat16,
    default_padding_side='left',
    device="cpu"
)
tokenizer = AutoTokenizer.from_pretrained(MODEL_TYPE)
tokenizer.padding_side = 'left'
tokenizer.pad_token = tokenizer.eos_token

**Comment by mylesgoose** [comment]
MODEL_PATH = 'meta-llama/Meta-Llama-3-8B-Instruct'

# little hack/tip:
# if you're dealing with a fine-tuned model of a "supported" model by transformer lens
# you can replicate the 'model path' of the supported model in your working directory
# e.g. rename the folder of 'dolphin-2.9-llama3-8b' to 'Meta-Llama-3-70B-Instruct', and put that into a folder called 'meta-llama'
# now transformers will accept 'meta-llama/Meta-Llama-3-70B-Instruct' as the model path for the model you're using, AND you don't have to add the model name to HookedTransformers
# make sure the model architecture and configs really do match though! 

model = HookedTransformer.from_pretrained_no_processing(
    MODEL_PATH,
    #local_files_only=True, # you can use local_files_only=True as a kwarg to from_pretrained_no_processing to enforce using the model from a local directory
    dtype=torch.bfloat16,   # you may want to try full precision if you can. bfloat16 is a good compromise though, but may not work in certain conditions or on certain hardware. DYOR
    default_padding_side='left'
)
https://huggingface.co/failspy/llama-3-70B-Instruct-abliterated/blob/main/ortho_cookbook.ipynb in this situation he is loading a renamed llama model of the same type however it does work with the llama3.1 model

**Comment by LeeSomm** [comment]
Hi @mylesgoose,
I'm interested in following the steps you outlined to make this work with Llama-3.1-8B, but I'm having difficulty understanding which files you're suggesting we edit and the formatting is a bit unclear. 

I understand replacing the loading_from_pretrained.py with the one you shared, but what are the other files you're referring to? 

Thanks in advance for any help!

**Comment by mylesgoose** [comment]
To resolve the issue using the simplest approach without editing any files, follow these steps:

Load LLaMA3 Version: Ensure that your script or environment is set up to use the LLaMA3 directory for initial loading. This is where your code will read the configurations and models.

Local Files Only: Confirm that your script or setup only uses local files and does not attempt to download or fetch any additional data from external sources.

Replace Files:

Once you have successfully loaded the LLaMA3 version, go to the LLaMA3 directory.
Delete all existing contents in the LLaMA3 directory.
Copy all contents from the LLaMA3.1 folder into the now-empty LLaMA3 directory.
Trick the Script: By replacing the contents in the LLaMA3 directory with those from the LLaMA3.1 folder, you will trick the script into thinking it’s loading the LLaMA3 version while actually using the tokenizer details and other configurations from the LLaMA3.1 version.

This method should allow your script to function as if it’s using LLaMA3, but with the updated details from LLaMA3.1. MODEL_PATH,
#local_files_only=True, # you can use local_files_only=True as a kwarg to from_pretrained_no_processing to enforce using the model from a local directory so test it with laodogin the llam 3 verison first yoru script. and if it works laodign form lcoal files then do above and kepe the file names in the scritp that for llama 3.0 for exampel `MODEL_ID = "meta-llama/Meta-Llama-3-70B-Instruct"
NEW_MODEL_ID = "mylesgoose/Meta-Llama-3.1-70B-abliterated"
MODEL_TYPE = "meta-llama/Meta-Llama-3-70B-Instruct"  as you can see the model type parameter will be shwoign to the scritp as llama 3 yet it will actually load the repalced llama 3.1 fiel you put intot he directory. ocne you tsted the lalam 3 one laoded.

**Comment by bryce13950** [comment]
@mylesgoose Are you interested in opening a PR, and adding the family officially? I am happy to do it, but you already did the work to support it, and I don't want to step on your toes when you could be credited officially on the repo.

**Comment by mylesgoose** [comment]
I think it already supported. the 70 b models and the 3.2 models. as I ran a script yesterday and checked inside the model template and it has been updated.  to include 3.2 models. I am currently working on tye vision  models. which managed ro get to work for the text model so far. on 3.2 11b. will do pr when finished. cheers

**Comment by bryce13950** [comment]
@mylesgoose 3.1 models are not currently supported. 3.2 are supported 3.1 are not. If you want to add them, I can put a release up quickly to add 3.1.

**Comment by mylesgoose** [comment]
okay I'll do it for you tomorrow.  I can see that the 3.1 8b and 70b models are not there. your right. the new ones where 3.2 1b and 3b . I have managed to load the 8b and 70b models 3.1 and hook them. I also managed to hook the vision model 11b 3.2 today but was outputting carbage but making progress. how can I modify the files here. do you want me to modify another repo then do pull requests

**Comment by bryce13950** [comment]
Yep! Are you on the open source slack? I can get you on there if you like, and we can setup a call today to go through the process. The vision models are probably going to be slightly more complicated. I am working with someone on LLaVA right now, and honestly we are not sure the way to proceed for adding different types of models to TransformerLens. It is an open debate at the moment, with myself preferring the idea of turning TransformerLens into a platform with a way to extend the base library for things like visions models, which would in turn keep TransformerLens itself purely focused on making interpretability of LLMs as good as it can be. It's an open debate though. For the time being, if you can get a vision model to load with minimal fuss, then I think we would be happy to add the support. It may take more than what it appears on the surface though. If you need help let me know!

**Comment by mylesgoose** [comment]
I made a couple of llava models. so a bit familiar  with them. well with regards to the open source slack. I don't know about it. with regards to the repo. the issue with the multi modal was I had to pass the cfg into hooketransfomer config and I tried for days to flatten the nested confg out. but as there is duplicated keys for text config and vision. models it was a nightmare. eventually  I settled for diverting hooked transfomer if it detected it was a nested text vision model in config for a new hookedVisiontransformerconfig class. so this what I'm working on tomorrow. and I actually  managed to get a new abliterated  model saved from  the  process but it only saved the text part of the model. so I have had to rethink and learn more. and I went into the core mllama transformers  library and found gold mine haha. it's pretty well explained how to do everything.  the only problem  with the way I'm doing it is if you want to add support  for other vision models you would then have to add a new clause into the hooked transformer config to pass that to the second llava visionhooked transfomer class. and I just passed the entire cfg from the model declaration page for the vision model to the htconfig. and Delt with it there. one confusing thing with the vision models is the weights change labels depending on how deep in the trasnfomer library it's getting. so for example  it changed from model.weights... to language_model.weights. vision_model... etc they then pass it again down another layer. quite confusing.

**Comment by mylesgoose** [comment]
> ### Proposal
> Add Llama 3.1 support. Currently trying to load it fails with:
> 
> `ValueError: meta-llama/Meta-Llama-3.1-8B-Instruct not found. Valid official model names (excl aliases): `
> 
> ### Motivation
> Llamma 3.1 is new bestest model, they say it's smarter than Rey Palpatine.
> 
> * [x]  I have checked that there is no similar [issue](https://github.com/TransformerLensOrg/Transformerlens/issues) in the repo (**required**)

https://github.com/mylesgoose/TransformerLens @ssuukk @bryce13950

**Comment by chigkim** [comment]
It would be great if TransformerLens supports vision models as well.
There are abliterated [recent vision models](https://huggingface.co/models?pipeline_tag=image-text-to-text&sort=created&search=abliterated), but I'm not sure how they  are done.
There seems to be a method to [remove refusals with just Transformers](https://github.com/Sumandora/remove-refusals-with-transformers/), but it didn't work with llama-3.2-vision.
I'd appreciate any tip to get llama-3.2-vision models to work with [FailSpy/abliterator](https://github.com/FailSpy/abliterator), or [Sumandora/remove-refusals-with-transformers](https://github.com/Sumandora/remove-refusals-with-transformers).
Thanks!

**Comment by chigkim** [comment]
Llama-3.2-vision model seems to come as one combined safetensors.
https://huggingface.co/meta-llama/Llama-3.2-11B-Vision-Instruct/tree/main
@mylesgoose Do you know  how to separate language and vision and combine  them together?

**Comment by chigkim** [comment]
Wow, thanks @mylesgoose for the script! I was successfully able to extract 8b language model from 11b vision language.
Now I need to find a script that can abliterate it.
I ran the script from [Sumandora/remove-refusals-with-transformers/](https://github.com/Sumandora/remove-refusals-with-transformers/issues/1) and saved into safetensor with no error.
However, when I tried to load the modified 8b language model with transformers and chat, it works like the original model and refuses non-safe requests.
I guess more puzzle to figure out, but thanks again!
